File: BaekJoon/src/shortestPath/BaekJoon_1504.java
package shortestPath;

import java.io.*;
import java.util.*;

/**
 * BaekJoon_1504, 특정한 최단 경로
 * @author kevin-Arpe
 * 
 * Sketch Idea
 *  1. 특정 두 정점을 지나야 하면서 1에서 출발해 N 정점으로 이동해야함
 *  2. 지나야하는 두 정점을 가는 와중에 N 정점을 도착했더라도 문제 조건에 의해 이전 방문 간선, 정점을 재 방문 가능하므로 무시
 *  3. 다익스트라를 사용하되 거쳐야 하는 두 정점의 순서를 바꾼 총 2가지 경로 중 최단 거리를 정답으로 출력
 *
 */
public class BaekJoon_1504 {
    static class Node implements Comparable<Node> {
        int e, w;
        Node next;
        
        Node(int e, int w) {
            this.e = e;
            this.w = w;
        }
        
        Node(int e, int w, Node next) {
            this.e = e;
            this.w = w;
            this.next = next;
        }

        @Override
        public int compareTo(Node o) {
            return this.w - o.w;
        }
    }
    
    static final int INF = Integer.MAX_VALUE >> 2;
    static int V, E;
    static Node[] adjList;
    
    static int n1, n2;
    
    static int dijkstra(int s, int e) {
        PriorityQueue<Node> pq = new PriorityQueue<>();
        pq.offer(new Node(s, 0));
        int[] dist = new int[V + 1];
        Arrays.fill(dist, INF);
        dist[s] = 0;
        boolean[] isVisited = new boolean[V + 1];
        
        while (!pq.isEmpty()) {
            Node curr = pq.poll();
            if (isVisited[curr.e]) continue; 
            isVisited[curr.e] = true; 
            
            for (Node next=adjList[curr.e]; next!=null; next=next.next) {
                if (isVisited[next.e] || dist[next.e] <= dist[curr.e] + next.w) continue;
                dist[next.e] = dist[curr.e] + next.w;
                
                pq.offer(new Node(next.e, dist[next.e]));
            }
        }
        return dist[e];
    }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        V = Integer.parseInt(st.nextToken());
        E = Integer.parseInt(st.nextToken());
        
        adjList = new Node[V + 1];
        for (int i=0; i<E; i++) {
            st = new StringTokenizer(br.readLine());
            int s = Integer.parseInt(st.nextToken());
            int e = Integer.parseInt(st.nextToken());
            int w = Integer.parseInt(st.nextToken());
            
            adjList[s] = new Node(e, w, adjList[s]);
            adjList[e] = new Node(s, w, adjList[e]);
        }
        
        st = new StringTokenizer(br.readLine());
        n1 = Integer.parseInt(st.nextToken());
        n2 = Integer.parseInt(st.nextToken());
        br.close();
        
        int toN1 = dijkstra(1, n1);
        int toN2 = dijkstra(1, n2);
        int btwN1N2 = dijkstra(n1, n2);
        int toEnd1 = dijkstra(n1, V);
        int toEnd2 = dijkstra(n2, V);
        
        
        int answer = INF;
        
        if (toN1 != INF && btwN1N2 != INF && toEnd2 != INF) {
            answer = Math.min(answer, toN1 + btwN1N2 + toEnd2);
        }
        if (toN2 != INF && btwN1N2 != INF && toEnd1 != INF) {
            answer = Math.min(answer, toN2 + btwN1N2 + toEnd1);
        }

        if (answer == INF) answer = -1;
        System.out.println(answer);
    }

}

File: BaekJoon/src/shortestPath/BaekJoon_1504.js
/**
 * BaekJoon_1504, 특정한 최단 경로
 *  - 문제 분류: 최단 경로 알고리즘, 다익스트라
 */
const solution = () => {
  /* 변수 관리 */
  const input = require('fs').readFileSync('./dev/stdin/1504.txt').toString().trim().split('\n');
  const [N, E] = input[0].split(' ').map(Number);
  const adjList = Array.from({ length: N + 1 }, () => []);
  const [A, B] = input[E + 1].split(' ').map(Number);
  const INF = Number.MAX_SAFE_INTEGER;

  for (let i = 1; i <= E; i++) {
    const [x, y, d] = input[i].split(' ').map(Number);
    adjList[x].push([y, d]);
    adjList[y].push([x, d]);
  }

  /* 메인 로직 */
  const distA = dijkstra(A);
  const distB = dijkstra(B);

  const answer = Math.min(distA[1] + distA[B] + distB[N], distB[1] + distB[A] + distA[N]);

  /* 정답 반환 */
  return answer >= INF ? -1 : answer;

  // 다익스트라
  function dijkstra(s) {
    const heap = Heap();
    const dist = Array.from({ length: N + 1 }, () => INF);
    heap.offer([s, 0]);
    dist[s] = 0;

    while (heap.size() > 0) {
      const [curr, acc] = heap.poll();
      if (acc > dist[curr]) continue;

      for (const [next, d] of adjList[curr]) {
        if (dist[next] > acc + d) {
          dist[next] = acc + d;
          heap.offer([next, dist[next]]);
        }
      }
    }

    return dist;
  }

  // 최소 힙
  function Heap() {
    const arr = [null];

    const size = () => {
      return arr.length - 1;
    };

    const swap = (i, j) => {
      const temp = arr[i];
      arr[i] = arr[j];
      arr[j] = temp;
    };

    const offer = (next) => {
      arr.push(next);
      let pointer = arr.length - 1;
      while (pointer > 1) {
        const parent = Math.floor(pointer / 2);
        if (arr[pointer][1] >= arr[parent][1]) break;
        swap(pointer, parent);
        pointer = parent;
      }
    };

    const poll = () => {
      if (arr.length === 1) return null;
      swap(1, arr.length - 1);
      const result = arr.pop();
      let pointer = 1;
      while (true) {
        const left = pointer * 2;
        const right = pointer * 2 + 1;
        let smallest = pointer;

        if (left < arr.length && arr[left][1] < arr[smallest][1]) {
          smallest = left;
        }

        if (right < arr.length && arr[right][1] < arr[smallest][1]) {
          smallest = right;
        }

        if (smallest === pointer) break;

        swap(pointer, smallest);
        pointer = smallest;
      }

      return result;
    };

    return {
      offer,
      poll,
      size,
    };
  }
};

console.log(solution());

